### PR TITLE
Add mediation debug within the tool using the micro-integrator

### PR DESCRIPTION
**Pull request by rosensilva**

### CODE REVIEW

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/serverdefinition/carbon.microIntegrator.definition.xml [hunk 1]
```diff
@@ -80,6 +80,14 @@
...
             context="server"
             default="false" />
+            
+    <property
+            id="esb.debug"
+            label="WSO2_DEBUG"
+            type="boolean"
+            context="runtime"
+            default="false" />
+
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/serverdefinition/carbon.microIntegrator.definition.xml]: Shall we fix formatting?

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/plugin.xml [hunk 2]
```diff
@@ -51,6 +51,12 @@
                class="org.wso2.developerstudio.eclipse.carbonserver44microei.wizard.ExportDistributionAndRunAction"
                     id="org.wso2.developerstudio.eclipse.carbonserver44microei.exportandrun">
            </action>
+           <action
+                    label="Export Project Artifacts and Debug"
+                    icon="icons/car.png"
```
Comment (dakshika) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/plugin.xml]: change the icon

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/register/product/servers/MicroIntegratorInstance.java [hunk 3]
```diff
@@ -233,4 +236,33 @@ public int getOffset() {
...
 
+	/**
+	 * This will set the mediation debug parameter to the micro-ei instance
+	 * 
+	 * @param setDebug the configuration for the mediation debug server parameter
+	 */
+	public void setDebugMode(boolean setDebug) {
+		GenericServer dl = (GenericServer) server.loadAdapter(GenericServer.class, null);
+		Map<String, String> temserverInstanceProperties = dl.getServerInstanceProperties();
+		temserverInstanceProperties.put("esb.debug", String.valueOf(setDebug));
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/register/product/servers/MicroIntegratorInstance.java]: Shall we make esb.debug a constant

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/register/product/servers/MicroIntegratorInstance.java [hunk 4]
```diff
@@ -233,4 +236,33 @@ public int getOffset() {
 		return offset;
 	}
 
+	/**
+	 * This will set the mediation debug parameter to the micro-ei instance
+	 * 
+	 * @param setDebug the configuration for the mediation debug server parameter
+	 */
+	public void setDebugMode(boolean setDebug) {
+		GenericServer dl = (GenericServer) server.loadAdapter(GenericServer.class, null);
```
Comment (dakshika) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/register/product/servers/MicroIntegratorInstance.java]: add proper name

Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/register/product/servers/MicroIntegratorInstance.java]: Shall we

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package org.wso2.developerstudio.eclipse.carbonserver44microei.wizard;
```
Comment (GDLMadushanka) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java]: shall we add the licence header

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+import org.wso2.developerstudio.eclipse.distribution.project.validator.ProjectList;
+import org.wso2.developerstudio.eclipse.logging.core.IDeveloperStudioLog;
+import org.wso2.developerstudio.eclipse.logging.core.Logger;
+import org.wso2.developerstudio.eclipse.maven.util.MavenUtils;
+import org.wso2.developerstudio.eclipse.platform.core.model.AbstractListDataProvider.ListData;
+import org.wso2.developerstudio.eclipse.platform.core.project.export.util.ExportUtil;
+import org.wso2.developerstudio.eclipse.platform.core.utils.Constants;
+import org.wso2.developerstudio.eclipse.utils.file.FileUtils;
+
+public class DistributionProjectExportAndDebugWizard extends Wizard implements IExportWizard {
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: Class comment please

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+	private IProject selectedProject;
+	private MavenProject parentPrj;
+	private static IDeveloperStudioLog log = Logger.getLog(Activator.PLUGIN_ID);
+	private boolean initError = false;
+	private Map<String, DependencyData> projectList = new HashMap<String, DependencyData>();
+	private Map<String, Dependency> dependencyMap = new HashMap<String, Dependency>();
+	private Map<String, String> serverRoleList = new HashMap<String, String>();
+	private ArtifactTypeMapping artifactTypeMapping = new ArtifactTypeMapping();
+
+	private static final String CARFileName = "TestCompositeApplication";
```
Comment (dakshika) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: add comment to explain why we need static name

Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: Shall we add a comment why CAR file name is static?

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndRunWizard.java [hunk 8]
```diff
@@ -213,8 +213,13 @@ public boolean performFinish() {
 			}
 			IResource carbonArchive = ExportUtil.buildCAppProject(selectedProject);
 			FileUtils.copy(carbonArchive.getLocation().toFile(), destFileName);
-			
-			//restart internal micro integrator profile
+
+			// set the mediation debug off if the server is configured with mediation debug
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndRunWizard.java]: Reset the profile?

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.debug.core.DebugPlugin;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.debug.core.ILaunchManager;
+import org.wso2.developerstudio.eclipse.carbonserver44microei.Activator;
+import org.wso2.developerstudio.eclipse.logging.core.IDeveloperStudioLog;
+import org.wso2.developerstudio.eclipse.logging.core.Logger;
+
+public class MediationDebugLauncher extends Thread {
```
Comment (abeykoon) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java]: Class level comment please

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+	private static IDeveloperStudioLog log = Logger.getLog(Activator.PLUGIN_ID);
+
+	public void run() {
+		try {
+			// Get the internal debug launcher from the launch manager
+			ILaunchManager launchManager = DebugPlugin.getDefault().getLaunchManager();
+			// Get the ESB Debug profile from the launch manager
+			ILaunchConfiguration debugESBLauchConfig = findLaunchConfigurationByName(launchManager, DEBUG_PROFILE_NAME);
+			// Launch the debug mode
+			debugESBLauchConfig.launch("debug", new NullProgressMonitor());
```
Comment (dakshika) [inline plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java]: Add null check

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/serverdefinition/carbon.microIntegrator.definition.xml (hunk 1 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/serverdefinition/carbon.microIntegrator.definition.xml]: fixed...

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java (hunk 6 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: fixed...

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java (hunk 7 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: fixed...

Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndDebugWizard.java]: fixed...

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndRunWizard.java (hunk 8 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/DistributionProjectExportAndRunWizard.java]: fixed...

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java (hunk 9 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java]: fixed...

File: plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java (hunk 10 above)
Comment (rosensilva) [reply plugins/org.wso2.developerstudio.eclipse.carbonserver44microei/src/org/wso2/developerstudio/eclipse/carbonserver44microei/wizard/MediationDebugLauncher.java]: fixed...

**Pull request merged**